tests/integration/conftest.py

```diff
@@ -1,16 +1,52 @@
 # Copyright 2021 Canonical Ltd.
 # See LICENSE file for licensing details.
+import functools
+import logging
 import shutil
+from collections import defaultdict
+from datetime import datetime
 
 import pytest
 from pytest_operator.plugin import OpsTest
 
+logger = logging.getLogger(__name__)
 
-@pytest.fixture(scope="module")
-async def loki_charm(ops_test: OpsTest):
-    """Loki charm used for integration testing."""
-    charm = await ops_test.build_charm(".")
-    return charm
+
+class Store(defaultdict):
+    def __init__(self):
+        super(Store, self).__init__(Store)
+
+    def __getattr__(self, key):
+        """Override __getattr__ so dot syntax works on keys."""
+        try:
+            return self[key]
+        except KeyError:
+            raise AttributeError(key)
+
+    def __setattr__(self, key, value):
+        """Override __setattr__ so dot syntax works on keys."""
+        self[key] = value
+
+
+store = Store()
+
+
+def timed_memoizer(func):
```

Comment:
This memoizer is not timed though, is it?

Reply:
[Yes, it is](https://github.com/canonical/loki-k8s-operator/pull/139/files#diff-c87c5173e88bb0f60b00107f4586304f4869a976a4f998425efe3481c0db9f56R46). It reports how long it took to either build (sometimes long) or return the memoized version (very fast). It's not `memoizer_with_timeout`. It's literally timing execution and logging it.